README.md
# Statistics plugin to Control tower api-gateway


CHANGELOG.md
# 2.0.0
- Updated license and authorship info
- Rename npm package to `sd-ct-statistics-plugin`


cron.js
require('app-module-path').addPath(__dirname);
const CronJob = require('cron').CronJob;
const mongoose = require('mongoose');
const debug = require('debug')('statistics-plugin');
const StatisticService = require('./lib/statistic.service');

module.exports = function cron(plugin, generalConfig) {
    debug('Loading statistics cron');
    const connection = mongoose.createConnection(`${generalConfig.mongoUri}`);
    const statisticService = new StatisticService(connection);
    async function tick() {
        try {
            debug('Executing tick in statistics microservice');
            await statisticService.completeGeoInfo();
        } catch (error) {
            debug('Error: ', error);
        }
    }

    new CronJob('00 * * * * *', tick, null, true, 'America/Los_Angeles'); // eslint-disable-line no-new
};


lib/stadistic.model.js
const mongoose = require('mongoose');
const Schema = mongoose.Schema;
const bluebird = require('bluebird');

mongoose.Promise = bluebird;

let stadistic = null;
function stadisticModel(connection) {
    if (stadistic) {
        return stadistic;
    }

    const Stadistic = new Schema({
        sourcePath: { type: String, required: true, trim: true },
        sourceMethod: { type: String, required: true, trim: true },
        redirectUrl: { type: String, required: false, trim: true },
        redirectMethod: { type: String, required: false, trim: true },
        endpointPath: { type: String, required: false, trim: true },
        time: { type: Number, required: true },
        date: { type: Date, required: true, default: Date.now },
        cached: { type: Boolean, required: true, default: false },
        error: { type: Boolean, required: true, default: false },
        errorCode: { type: Number, required: false },
        ip: { type: String, required: false, trim: true },
        anonymous: { type: Boolean, required: true, default: true },
        loggedUser: { type: Schema.Types.Mixed },
        geo: {
            _id: false,
            city: { type: String, required: false, trim: true },
            country: { type: String, required: false, trim: true },
            region: { type: String, required: false, trim: true },
            ll: [{ type: Number, required: false, trim: true }],
            completed: { type: Boolean, required: true, default: false },
        },
    });
    stadistic = connection.model('Stadistic', Stadistic);
    return stadistic;
}

module.exports = stadisticModel;


lib/statistic.model.js
const mongoose = require('mongoose');
const Schema = mongoose.Schema;
const bluebird = require('bluebird');

mongoose.Promise = bluebird;

let statistic = null;
function statisticModel(connection) {
    if (statistic) {
        return statistic;
    }

    const Statistic = new Schema({
        sourcePath: { type: String, required: true, trim: true },
        sourceMethod: { type: String, required: true, trim: true },
        redirectUrl: { type: String, required: false, trim: true },
        redirectMethod: { type: String, required: false, trim: true },
        endpointPath: { type: String, required: false, trim: true },
        time: { type: Number, required: true },
        date: { type: Date, required: true, default: Date.now },
        cached: { type: Boolean, required: true, default: false },
        error: { type: Boolean, required: true, default: false },
        errorCode: { type: Number, required: false },
        ip: { type: String, required: false, trim: true },
        anonymous: { type: Boolean, required: true, default: true },
        loggedUser: { type: Schema.Types.Mixed },
        geo: {
            _id: false,
            city: { type: String, required: false, trim: true },
            country: { type: String, required: false, trim: true },
            region: { type: String, required: false, trim: true },
            ll: [{ type: Number, required: false, trim: true }],
            completed: { type: Boolean, required: true, default: false },
        },
    });
    statistic = connection.model('Statistic', Statistic);
    return statistic;
}

module.exports = statisticModel;


lib/stadistic.router.js
const debug = require('debug')('stadistics-plugin');
const Router = require('koa-router');
const stadisticFunction = require('./stadistic.model');
const ApiRouter = new Router({
    prefix: '/api/v1/stadistic',
});


function getMiddleware(connection) {
    const StadisticModel = stadisticFunction(connection);
    class StadisticRouter {

        static async get(ctx) {
            debug('Obtaining stadistics');
            ctx.body = await StadisticModel.find().sort('-date').exec();
        }

        static async timeByRequest(ctx) {
            debug('Obtaining stadistics aggrouped');
            let filter = null;
            debug('start', ctx.query.from);
            if (ctx.query.from || ctx.query.to) {
                filter = {
                    $match: {
                        date: {},
                    },
                };
                if (ctx.query.from) {
                    filter.$match.date.$gte = new Date(new Date(ctx.query.from).getTime() - (new Date(ctx.query.from).getTimezoneOffset() * 60000));
                }
                if (ctx.query.to) {
                    filter.$match.date.$lte = new Date(new Date(ctx.query.to).getTime() - (new Date(ctx.query.to).getTimezoneOffset() * 60000) + (24 * 60 * 60 * 1000) - 60000);
                }
            }
            const query = [];
            if (filter) {
                query.push(filter);
            }
            query.push({
                $group: {
                    _id: {
                        endpointPath: '$endpointPath',
                        sourceMethod: '$sourceMethod',
                    },
                    sum: {
                        $sum: '$time',
                    },

                },
            });
            ctx.body = await StadisticModel.aggregate(query).exec();
        }

        static async avgByRequest(ctx) {
            debug('Obtaining stadistics aggrouped');
            let filter = null;
            debug('start', ctx.query.from);
            if (ctx.query.from || ctx.query.to) {
                filter = {
                    $match: {
                        date: {},
                    },
                };
                if (ctx.query.from) {
                    filter.$match.date.$gte = new Date(new Date(ctx.query.from).getTime() - (new Date(ctx.query.from).getTimezoneOffset() * 60000));
                }
                if (ctx.query.to) {
                    filter.$match.date.$lte = new Date(new Date(ctx.query.to).getTime() - (new Date(ctx.query.to).getTimezoneOffset() * 60000) + (24 * 60 * 60 * 1000) - 60000);
                }
            }
            const query = [];
            if (filter) {
                query.push(filter);
            }
            query.push({
                $group: {
                    _id: {
                        endpointPath: '$endpointPath',
                        sourceMethod: '$sourceMethod',
                    },
                    sum: {
                        $avg: '$time',
                    },
                    count: { $sum: 1 },
                },
            });
            ctx.body = await StadisticModel.aggregate(query).exec();
        }

        static async countRequestToday(ctx) {
            debug('Obtained request of today');
            const date = new Date();
            const today = new Date(date.getFullYear(), date.getMonth(), date.getDate(), 0, 0, 0);
            const tomorrowDate = new Date(Date.now() + 24 * 60 * 60 * 1000);
            const tomorrow = new Date(tomorrowDate.getFullYear(), tomorrowDate.getMonth(), tomorrowDate.getDate(), 0, 0, 0);

            const count = await StadisticModel.count({
                date: {
                    $lt: tomorrow,
                    $gte: today,
                },
            }).exec();
            ctx.body = {
                count,
                begin: today.toISOString(),
                end: tomorrow.toISOString(),
            };
        }

        static async countRequestLastWeek(ctx) {
            debug('Obtained request of today');
            const lastWeekDate = new Date(Date.now() - 7 * 24 * 60 * 60 * 1000);
            const tomorrowDate = new Date(Date.now() + 24 * 60 * 60 * 1000);
            const lastWeek = new Date(lastWeekDate.getFullYear(), lastWeekDate.getMonth(), lastWeekDate.getDate(), 0, 0, 0);
            const tomorrow = new Date(tomorrowDate.getFullYear(), tomorrowDate.getMonth(), tomorrowDate.getDate(), 0, 0, 0);


            const count = await StadisticModel.count({
                date: {
                    $lt: tomorrow,
                    $gte: lastWeek,
                },
            }).exec();

            ctx.body = {
                count,
                begin: lastWeek.toISOString(),
                end: tomorrow.toISOString(),
            };
        }

        static async countRequestTodayByCountry(ctx) {
            debug('Obtaining num request today per country');
            const date = new Date();
            const today = new Date(date.getFullYear(), date.getMonth(), date.getDate(), 0, 0, 0);
            const tomorrowDate = new Date(Date.now() + 24 * 60 * 60 * 1000);
            const tomorrow = new Date(tomorrowDate.getFullYear(), tomorrowDate.getMonth(), tomorrowDate.getDate(), 0, 0, 0);

            const result = await StadisticModel.aggregate([{
                $match: {
                    date: {
                        $lt: tomorrow,
                        $gte: today,
                    },
                },
            }, {
                $group: {
                    _id: '$geo.country',
                    count: {
                        $sum: 1,
                    },
                },
            }]).exec();
            ctx.body = result;
        }

        static async requestByDay(ctx) {
            debug('Obtaining stadistics aggrouped');
            let filter = null;
            if (ctx.query.from || ctx.query.to) {
                filter = {
                    $match: {
                        date: {},
                    },
                };
                if (ctx.query.from) {
                    filter.$match.date.$gte = new Date(new Date(ctx.query.from).getTime() - (new Date(ctx.query.from).getTimezoneOffset() * 60000));
                }
                if (ctx.query.to) {
                    filter.$match.date.$lte = new Date(new Date(ctx.query.to).getTime() - (new Date(ctx.query.to).getTimezoneOffset() * 60000) + (24 * 60 * 60 * 1000) - 60000);
                }
            }
            const query = [];
            if (filter) {
                query.push(filter);
            }
            query.push({
                $group: {
                    _id: {
                        year: {
                            $year: '$date',
                        },
                        month: {
                            $month: '$date',
                        },
                        day: {
                            $dayOfMonth: '$date',
                        },
                    },
                    count: {
                        $sum: 1,
                    },
                },
            });
            ctx.body = await StadisticModel.aggregate(query).exec();
        }

    }

    ApiRouter.get('/requestByDay', StadisticRouter.requestByDay);
    ApiRouter.get('/countRequestTodayByCountry', StadisticRouter.countRequestTodayByCountry);
    ApiRouter.get('/countRequestToday', StadisticRouter.countRequestToday);
    ApiRouter.get('/countRequestLastWeek', StadisticRouter.countRequestLastWeek);
    ApiRouter.get('/timeByRequest', StadisticRouter.timeByRequest);
    ApiRouter.get('/avgByRequest', StadisticRouter.avgByRequest);
    ApiRouter.get('/', StadisticRouter.get);

    return ApiRouter;
}
module.exports = getMiddleware;


lib/stadistic.service.js
const debug = require('debug')('stadistics-plugin');
const stadisticFunction = require('./stadistic.model');
const geoip = require('geoip-lite');

function getService(connection) {

    const StadisticModel = stadisticFunction(connection);

    class StadisticService {

        static async completeGeoInfo() {
            debug('Loading stadistics to complete geo info');
            const stadistics = await StadisticModel.find({
                ip: {
                    $exists: true,
                },
                'geo.completed': false,
            }).limit(10000).exec();
            debug('Ips found ', stadistics.length);
            for (let i = 0, length = stadistics.length; i < length; i++) {
                if (stadistics[i].ip && stadistics[i].ip.indexOf('127.0.0.1') === -1) {
                    let ip = stadistics[i].ip;
                    if (ip.indexOf(',') >= 0) {
                        ip = ip.split(',')[1];
                    }
                    const geo = geoip.lookup(ip);
                    if (geo) {
                        stadistics[i].geo = {
                            city: geo.city,
                            country: geo.country,
                            region: geo.region,
                            ll: geo.ll,
                            completed: true,
                        };
                    } else {
                        stadistics[i].geo = {
                            completed: true,
                        };
                    }
                } else {
                    stadistics[i].geo = {
                        completed: true,
                    };
                }
                await stadistics[i].save();
            }
            debug('Finish complete geo');
        }

        static async middleware(ctx, next) {
            const first = Date.now();
            let error = false;
            let errorCode = null;
            try {
                await next();
            } catch (e) {
                error = true;
                errorCode = e.status || 500;
                throw e;
            } finally {
                if (ctx.state.source && ctx.state.source.path) {
                    const model = {
                        sourcePath: ctx.state.source.path,
                        sourceMethod: ctx.state.source.method,
                        error,
                        errorCode,
                        cached: false,
                        time: Date.now() - first,
                        ip: ctx.headers['x-forwarded-for'],
                        anonymous: (!ctx.state.user && !ctx.req.user && !ctx.state.microservice),
                        loggedUser: ctx.state.user || ctx.req.user || ctx.state.microservice,
                    };
                    if (ctx.state.redirect) {
                        model.endpointPath = ctx.state.redirect.endpoint.path;
                        model.redirectUrl = ctx.state.redirect.url;
                        model.redirectMethod = ctx.state.redirect.method;
                    }

                    debug('Saving stadistic');
                    await new StadisticModel(model).save();
                } else {
                    const model = {
                        sourcePath: ctx.path,
                        sourceMethod: ctx.request.method,
                        error,
                        errorCode,
                        cached: ctx.state.isCached || false,
                        time: Date.now() - first,
                        ip: ctx.headers['x-forwarded-for'],
                        anonymous: (!ctx.state.user && !ctx.req.user && !ctx.state.microservice),
                        loggedUser: ctx.state.user || ctx.req.user || ctx.state.microservice
                    };
                    if (ctx.state.redirect) {
                        model.endpointPath = ctx.state.redirect.endpoint.path;
                        model.redirectUrl = ctx.state.redirect.url;
                        model.redirectMethod = ctx.state.redirect.method;
                    }


                    debug('Saving stadistic');
                    await new StadisticModel(model).save();

                }
            }
        }

    }
    return StadisticService;
}
module.exports = getService;


lib/statistic.service.js
const debug = require('debug')('statistics-plugin');
const statisticFunction = require('./statistic.model');
const geoip = require('geoip-lite');

function getService(connection) {

    const StatisticModel = statisticFunction(connection);

    class StatisticService {

        static async completeGeoInfo() {
            debug('Loading statistics to complete geo info');
            const statistics = await StatisticModel.find({
                ip: {
                    $exists: true,
                },
                'geo.completed': false,
            }).limit(10000).exec();
            debug('Ips found ', statistics.length);
            for (let i = 0, length = statistics.length; i < length; i++) {
                if (statistics[i].ip && statistics[i].ip.indexOf('127.0.0.1') === -1) {
                    let ip = statistics[i].ip;
                    if (ip.indexOf(',') >= 0) {
                        ip = ip.split(',')[1];
                    }
                    const geo = geoip.lookup(ip);
                    if (geo) {
                        statistics[i].geo = {
                            city: geo.city,
                            country: geo.country,
                            region: geo.region,
                            ll: geo.ll,
                            completed: true,
                        };
                    } else {
                        statistics[i].geo = {
                            completed: true,
                        };
                    }
                } else {
                    statistics[i].geo = {
                        completed: true,
                    };
                }
                await statistics[i].save();
            }
            debug('Finish complete geo');
        }

        static async middleware(ctx, next) {
            const first = Date.now();
            let error = false;
            let errorCode = null;
            try {
                await next();
            } catch (e) {
                error = true;
                errorCode = e.status || 500;
                throw e;
            } finally {
                if (ctx.state.source && ctx.state.source.path) {
                    const model = {
                        sourcePath: ctx.state.source.path,
                        sourceMethod: ctx.state.source.method,
                        error,
                        errorCode,
                        cached: false,
                        time: Date.now() - first,
                        ip: ctx.headers['x-forwarded-for'],
                        anonymous: (!ctx.state.user && !ctx.req.user && !ctx.state.microservice),
                        loggedUser: ctx.state.user || ctx.req.user || ctx.state.microservice,
                    };
                    if (ctx.state.redirect) {
                        model.endpointPath = ctx.state.redirect.endpoint.path;
                        model.redirectUrl = ctx.state.redirect.url;
                        model.redirectMethod = ctx.state.redirect.method;
                    }

                    debug('Saving statistic');
                    await new StatisticModel(model).save();
                } else {
                    const model = {
                        sourcePath: ctx.path,
                        sourceMethod: ctx.request.method,
                        error,
                        errorCode,
                        cached: ctx.state.isCached || false,
                        time: Date.now() - first,
                        ip: ctx.headers['x-forwarded-for'],
                        anonymous: (!ctx.state.user && !ctx.req.user && !ctx.state.microservice),
                        loggedUser: ctx.state.user || ctx.req.user || ctx.state.microservice
                    };
                    if (ctx.state.redirect) {
                        model.endpointPath = ctx.state.redirect.endpoint.path;
                        model.redirectUrl = ctx.state.redirect.url;
                        model.redirectMethod = ctx.state.redirect.method;
                    }


                    debug('Saving statistic');
                    await new StatisticModel(model).save();

                }
            }
        }

    }
    return StatisticService;
}
module.exports = getService;


lib/statistic.router.js
const debug = require('debug')('statistics-plugin');
const Router = require('koa-router');
const statisticFunction = require('./statistic.model');
const ApiRouter = new Router({
    prefix: '/api/v1/statistic',
});


function getMiddleware(connection) {
    const StatisticModel = statisticFunction(connection);
    class StatisticRouter {

        static async get(ctx) {
            debug('Obtaining statistics');
            ctx.body = await StatisticModel.find().sort('-date').exec();
        }

        static async timeByRequest(ctx) {
            debug('Obtaining grouped statistics');
            let filter = null;
            debug('start', ctx.query.from);
            if (ctx.query.from || ctx.query.to) {
                filter = {
                    $match: {
                        date: {},
                    },
                };
                if (ctx.query.from) {
                    filter.$match.date.$gte = new Date(new Date(ctx.query.from).getTime() - (new Date(ctx.query.from).getTimezoneOffset() * 60000));
                }
                if (ctx.query.to) {
                    filter.$match.date.$lte = new Date(new Date(ctx.query.to).getTime() - (new Date(ctx.query.to).getTimezoneOffset() * 60000) + (24 * 60 * 60 * 1000) - 60000);
                }
            }
            const query = [];
            if (filter) {
                query.push(filter);
            }
            query.push({
                $group: {
                    _id: {
                        endpointPath: '$endpointPath',
                        sourceMethod: '$sourceMethod',
                    },
                    sum: {
                        $sum: '$time',
                    },

                },
            });
            ctx.body = await StatisticModel.aggregate(query).exec();
        }

        static async avgByRequest(ctx) {
            debug('Obtaining statistics aggrouped');
            let filter = null;
            debug('start', ctx.query.from);
            if (ctx.query.from || ctx.query.to) {
                filter = {
                    $match: {
                        date: {},
                    },
                };
                if (ctx.query.from) {
                    filter.$match.date.$gte = new Date(new Date(ctx.query.from).getTime() - (new Date(ctx.query.from).getTimezoneOffset() * 60000));
                }
                if (ctx.query.to) {
                    filter.$match.date.$lte = new Date(new Date(ctx.query.to).getTime() - (new Date(ctx.query.to).getTimezoneOffset() * 60000) + (24 * 60 * 60 * 1000) - 60000);
                }
            }
            const query = [];
            if (filter) {
                query.push(filter);
            }
            query.push({
                $group: {
                    _id: {
                        endpointPath: '$endpointPath',
                        sourceMethod: '$sourceMethod',
                    },
                    sum: {
                        $avg: '$time',
                    },
                    count: { $sum: 1 },
                },
            });
            ctx.body = await StatisticModel.aggregate(query).exec();
        }

        static async countRequestToday(ctx) {
            debug('Obtained request of today');
            const date = new Date();
            const today = new Date(date.getFullYear(), date.getMonth(), date.getDate(), 0, 0, 0);
            const tomorrowDate = new Date(Date.now() + 24 * 60 * 60 * 1000);
            const tomorrow = new Date(tomorrowDate.getFullYear(), tomorrowDate.getMonth(), tomorrowDate.getDate(), 0, 0, 0);

            const count = await StatisticModel.count({
                date: {
                    $lt: tomorrow,
                    $gte: today,
                },
            }).exec();
            ctx.body = {
                count,
                begin: today.toISOString(),
                end: tomorrow.toISOString(),
            };
        }

        static async countRequestLastWeek(ctx) {
            debug('Obtained request of today');
            const lastWeekDate = new Date(Date.now() - 7 * 24 * 60 * 60 * 1000);
            const tomorrowDate = new Date(Date.now() + 24 * 60 * 60 * 1000);
            const lastWeek = new Date(lastWeekDate.getFullYear(), lastWeekDate.getMonth(), lastWeekDate.getDate(), 0, 0, 0);
            const tomorrow = new Date(tomorrowDate.getFullYear(), tomorrowDate.getMonth(), tomorrowDate.getDate(), 0, 0, 0);


            const count = await StatisticModel.count({
                date: {
                    $lt: tomorrow,
                    $gte: lastWeek,
                },
            }).exec();

            ctx.body = {
                count,
                begin: lastWeek.toISOString(),
                end: tomorrow.toISOString(),
            };
        }

        static async countRequestTodayByCountry(ctx) {
            debug('Obtaining num request today per country');
            const date = new Date();
            const today = new Date(date.getFullYear(), date.getMonth(), date.getDate(), 0, 0, 0);
            const tomorrowDate = new Date(Date.now() + 24 * 60 * 60 * 1000);
            const tomorrow = new Date(tomorrowDate.getFullYear(), tomorrowDate.getMonth(), tomorrowDate.getDate(), 0, 0, 0);

            const result = await StatisticModel.aggregate([{
                $match: {
                    date: {
                        $lt: tomorrow,
                        $gte: today,
                    },
                },
            }, {
                $group: {
                    _id: '$geo.country',
                    count: {
                        $sum: 1,
                    },
                },
            }]).exec();
            ctx.body = result;
        }

        static async requestByDay(ctx) {
            debug('Obtaining grouped statistics');
            let filter = null;
            if (ctx.query.from || ctx.query.to) {
                filter = {
                    $match: {
                        date: {},
                    },
                };
                if (ctx.query.from) {
                    filter.$match.date.$gte = new Date(new Date(ctx.query.from).getTime() - (new Date(ctx.query.from).getTimezoneOffset() * 60000));
                }
                if (ctx.query.to) {
                    filter.$match.date.$lte = new Date(new Date(ctx.query.to).getTime() - (new Date(ctx.query.to).getTimezoneOffset() * 60000) + (24 * 60 * 60 * 1000) - 60000);
                }
            }
            const query = [];
            if (filter) {
                query.push(filter);
            }
            query.push({
                $group: {
                    _id: {
                        year: {
                            $year: '$date',
                        },
                        month: {
                            $month: '$date',
                        },
                        day: {
                            $dayOfMonth: '$date',
                        },
                    },
                    count: {
                        $sum: 1,
                    },
                },
            });
            ctx.body = await StatisticModel.aggregate(query).exec();
        }

    }

    ApiRouter.get('/requestByDay', StatisticRouter.requestByDay);
    ApiRouter.get('/countRequestTodayByCountry', StatisticRouter.countRequestTodayByCountry);
    ApiRouter.get('/countRequestToday', StatisticRouter.countRequestToday);
    ApiRouter.get('/countRequestLastWeek', StatisticRouter.countRequestLastWeek);
    ApiRouter.get('/timeByRequest', StatisticRouter.timeByRequest);
    ApiRouter.get('/avgByRequest', StatisticRouter.avgByRequest);
    ApiRouter.get('/', StatisticRouter.get);

    return ApiRouter;
}
module.exports = getMiddleware;
